### PR TITLE
Track a "human sequence number" for keeps

### PR DESCRIPTION
**Pull request by ryanpbrewster**
@leogrim

image

### CODE REVIEW

File: kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala [hunk 1]
```diff
@@ -219,6 +225,7 @@ class KeepRepoImpl @Inject() (
       val privateVisibilities: Set[LibraryVisibility] = Set(LibraryVisibility.SECRET, LibraryVisibility.ORGANIZATION)
       visibility.inSet(privateVisibilities)
     }
+    def hseq: Column[SequenceNumber[Keep]] = deferredHseq.ifNull(seq)
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala:228]: I have no idea what this function does

File: kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala [hunk 2]
```diff
@@ -188,6 +193,7 @@ class KeepRepoImpl @Inject() (
 
   type RepoImpl = KeepTable
   class KeepTable(tag: Tag) extends RepoTable[Keep](db, tag, "bookmark") with ExternalIdColumn[Keep] with SeqNumberColumn[Keep] with NamedColumns {
+    def deferredHseq = column[Option[SequenceNumber[Keep]]]("hseq", O.Nullable) //indexd
```
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala:196]: Missing an index?

**Comment by andrewconner** [comment]
Mind summarizing a bit what this is for?

**Comment by ryanpbrewster** [comment]
A sequence number tracks whenever ANY change is made to a model.

A subsequence number tracks when a specific subset of changes are made to a model. In this case, a human sequence number tracks only human-initiated changes (no URI renormalization or integrity checking). This allows us to find the subset of keeps that have been changed in "interesting" ways.

In practice, if there is a human-initiated change made to a model, we wipe out it's "hseq" and the next time it is read in from memory the current `seq` number is slotting into place. If a system-initiated change is made, nothing happens and the `hseq` persists with the existing value, slowly falling behind the `seq` (as intended).

**Comment by andrewconner** [comment]
Is this so we can push updates somewhere if titles are updated manually?

**Comment by ryanpbrewster** [comment]
More or less.

File: kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala (hunk 1 above)
Comment (camhashemi) [inline kifi-backend/modules/shoebox/app/com/keepit/model/KeepRepo.scala:228]: lol

**Pull request closed without merging**